elyra/pipeline/component_catalog.py

```diff
@@ -195,36 +315,75 @@ class ComponentCache(SingletonConfigurable):
         super().__init__(**kwargs)
 
-        self._cache_task_queue = Queue()
+        self._cache_queue = Queue()
+        self._manifest_queue = Queue()
+
+        manifest_dir = kwargs.get('component_manifest_dir', jupyter_core.paths.jupyter_runtime_dir())
+        self.manifest_filename = os.path.join(manifest_dir, "elyra-component-manifest.json")
 
-        # Start a thread to manage updates to the component cache
-        self._cache_manager = CacheUpdateManagerThread(self.log, self._component_cache, self._cache_task_queue)
+        # Set up watchdog for manifest
+        self.observer = Observer()
+        self.observer.schedule(ManifestFileChangeHandler(self), manifest_dir)
+
+        # Start a thread to manage updates to the component cache and manifest
+        self._cache_manager = CacheUpdateManagerThread(
+            self.log,
+            self._component_cache,
+            self._cache_queue,
+            self._manifest_queue,
+            self.manifest_filename
+        )
         self._cache_manager.name = "CacheUpdateManager"
         self._cache_manager.start()
 
-        self._build_cache()
-
-    def _build_cache(self):
+    def load(self):
+        """
+        Completes a series of actions during system startup, such as creating
+        the component manifest file and triggering the build of the component
+        cache for existing ComponentCatalog metadata instances.
         """
-        Reads through all ComponentCatalogMetadata instances to build the initial component cache
+        # Proceed only if singleton instance has been created
+        if self.initialized:
+            # Create/overwrite manifest file and start its watchdog
+            self.set_up_manifest_file()
+            self.observer.start()
+
+            # Fetch all component catalog instances and trigger their add to
+            # the component manifest, and consequently, the component cache
+            catalogs = MetadataManager(schemaspace=ComponentCatalogs.COMPONENT_CATALOGS_SCHEMASPACE_ID).get_all()
+            for catalog in catalogs:
+                self.update_manifest(catalog, 'modify')
+
+    def set_up_manifest_file(self):
         """
+        Build an empty manifest and overwrite or create the file. After this point,
+        only the CacheUpdateManagerThread is able to touch this file.
+        """
+        empty_manifest = {
+            "actions": [],
+            "status": {}
+        }
 
-        all_catalogs = MetadataManager(schemaspace=ComponentCatalogs.COMPONENT_CATALOGS_SCHEMASPACE_ID).get_all()
-        for catalog in all_catalogs:
-            self.update_cache_for_catalog(catalog)
+        with open(self.manifest_filename, 'w') as f:
+            f.write(json.dumps(empty_manifest, indent=2))
 
-    def update_cache_for_catalog(self, catalog: ComponentCatalogMetadata, operation: Optional[str] = None):
+    def update_manifest(self, catalog: Optional[ComponentCatalogMetadata] = None, action: Optional[str] = None):
```

Comment:
Are these parameters optional?  I don't see any `None` handling in the related methods.

Reply:
Per offline discussion, `action` is never `None`, but catalog in `None` in the case where the watchdog is trigger an action. Will move `action` to the first parameter position and remove the 'Optional' hint.